src/backend/distributed/metadata/metadata_cache.c

```diff
@@ -4752,20 +4749,26 @@ InvalidateForeignKeyGraph(void)
  * InvalidateDistRelationCacheCallback flushes cache entries when a relation
  * is updated (or flushes the entire cache).
  */
-static void
+void
 InvalidateDistRelationCacheCallback(Datum argument, Oid relationId)
 {
 	/* invalidate either entire cache or a specific entry */
 	if (relationId == InvalidOid)
 	{
 		InvalidateDistTableCache();
 		InvalidateDistObjectCache();
+		InvalidateMetadataSystemCache();
 	}
 	else
 	{
 		void *hashKey = (void *) &relationId;
 		bool foundInCache = false;
 
+		if (DistTableCacheHash == NULL)
+		{
+			return;
+		}
+
 		CitusTableCacheEntrySlot *cacheSlot =
 			hash_search(DistTableCacheHash, hashKey, HASH_FIND, &foundInCache);
 		if (foundInCache)
```

Comment:
I think we can remove the InvalidateMetadataSystemCache call below here.

Reply:
Removing InvalidateMetadataSystemCache call here causes `check-vanilla` test suit to fail. Specifically ` check_index` is failing consistently in the CI if I remove this code. 

Otherwise, with my latest changes CI run succeeds although the results are not reflected in the "Checks" view here.

Reply:
It turns out that

REINDEX SCHEMA CONCURRENTLY pg_catalog;

command causes the relcache invalidation of pg_dist tables which are under pg_catalog. If we do not invalidate Metadatacache, it will end up with stale oids causing the subsequent commands to fail.

For instance, the following command
\d

will run an SELECT that invokes citus hooks and will cause failed cached lookups of pg_dist* tables.

So we need to invalidate the cache when one of the pg_dist* tables got invalidated.

Reply:
I have restored the previous code not to regress check-vanilla.
We will track the  issue of pg_dist_ tables getting invalidated in a seperate PR.

Note that, even if I restore the original code there are likely two issues with the original code
  1. When a pg_dist table other than `pg_dist_partition` gets invalidated for example using
        REINDEX TABLE pg_dist_background_task;
     we currently do not refresh its oid in the Metadata cache.

  2. When `pg_dist_partition` gets invalidated we unnecessarily nuke the entire MetadataCache.